File: app/src/main/java/com/atakaice/features/news/NewsCardsFragment.java
package com.atakaice.features.news;


import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.atakaice.R;
import com.atakaice.commons.News;
import com.atakaice.commons.NewsItem;
import com.atakaice.commons.RxBaseFragment;
import com.atakaice.features.news.adapter.ShuffleAdapter;
import com.meetic.shuffle.Shuffle;
import rx.Scheduler;
import rx.Subscription;
import rx.functions.Action1;
import rx.schedulers.Schedulers;

import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class NewsCardsFragment extends RxBaseFragment {

    private final String KEY_NEWS = "news";
    private News apiNews;
    private NewsManager newsManager;
    private Shuffle shuffle;

    public NewsCardsFragment() {
        // Required empty public constructor
        apiNews = null;
        newsManager = new NewsManager();
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_news_cards, container, false);
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);


        //setup adapter
        shuffle = getView().findViewById(R.id.shuffle);
        shuffle.setShuffleAdapter(new ShuffleAdapter());

        initAdapter();

        if(savedInstanceState != null && savedInstanceState.containsKey(KEY_NEWS)) {
            apiNews = (News)savedInstanceState.get(KEY_NEWS);
            ((ShuffleAdapter)shuffle.getShuffleAdapter()).clearAndAddNews(apiNews.getNews());
        } else {
            requestNews();
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        List<NewsItem> newsItems = ((ShuffleAdapter)shuffle.getShuffleAdapter()).getNews();
        if(apiNews != null && !newsItems.isEmpty()) {
            outState.putParcelable(KEY_NEWS, (Parcelable) apiNews);
        }

    }

    public void requestNews() {
        final String after = apiNews == null ? "" : apiNews.getAfter();
        Subscription subscription = newsManager.getNews(after, "10")
               .subscribeOn(Schedulers.io())
               .subscribe(
                       new Action1<News>() {
                           @Override
                           public void call(News retrivedNews) {
                               apiNews = retrivedNews;
                               ((ShuffleAdapter)shuffle.getShuffleAdapter()).addNews(retrivedNews.getNews());
                           }
                       });

        this.getSubscriptions().add(subscription);
    }

    private void initAdapter() {
        if(shuffle.getShuffleAdapter() == null) {
            shuffle.setShuffleAdapter(new ShuffleAdapter());
        }
    }
}

File: app/src/main/java/com/atakaice/features/news_cards/NewsCardsFragment.kt
package com.atakaice.features.news_cards

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.atakaice.R
import com.atakaice.commons.InfiniteScrollListener
import com.atakaice.commons.News
import com.atakaice.commons.RxBaseFragment
import com.atakaice.commons.extensions.inflate
import com.atakaice.features.news.NewsManager
import kotlinx.android.synthetic.main.fragment_news.*
import rx.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_news_cards.*


class NewsCardsFragment : RxBaseFragment() {

    companion object {
        private val KEY_NEWS = "news"
    }

    private var apiNews: News? = null
    private val newsManager by lazy { NewsManager() }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return container?.inflate(R.layout.fragment_news_cards)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        swipeStack.apply {
            var linearLayout = LinearLayoutManager(context)
        }

        initAdapter()

        if (savedInstanceState != null && savedInstanceState.containsKey(KEY_NEWS)) {
            apiNews = savedInstanceState.get(KEY_NEWS) as News
            //(news_list as NewsCardsAdapter).clearAndAddNews(apiNews!!.news)
            swipeStack.adapter = SwipeStackAdapter(context, apiNews!!.news)
        } else {
            requestNews()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val news = (swipeStack.adapter as SwipeStackAdapter).data
        if(apiNews != null && news.isNotEmpty()) {
            outState.putParcelable(KEY_NEWS, apiNews?.copy(news = news))
        }
    }

    private fun requestNews() {
        // (news_list.adapter as NewsAdapter).addNews(news)
        val subscription = newsManager.getNews(apiNews?.after ?: "", "10")
            .subscribeOn(Schedulers.io())
            .subscribe(
                {retrievedNews ->
                    apiNews = retrievedNews
                    //(news_list.adapter as NewsCardsAdapter).addNews(retrievedNews.news)
                    swipeStack.adapter = SwipeStackAdapter(context, retrievedNews.news)
                },
                {e -> Snackbar.make(swipeStack, e.message ?: "", Snackbar.LENGTH_LONG).show()
                Log.e("CONNECTION", e.message ?: "")}
            )

        subscriptions.add(subscription)
    }

    private fun initAdapter() {
        if (swipeStack.adapter == null) {
            swipeStack.adapter = SwipeStackAdapter(context)
        }
    }
}